plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -1013,6 +1024,122 @@ public void rollback()
+            ConnectorSession session,
+            ConnectorTableHandle handle,
+            List<ConnectorExpression> projections,
+            Map<String, ColumnHandle> assignments)
+    {
+        if (!isProjectionPushdownEnabled(session)) {
+            return Optional.empty();
+        }
+
+        // Create projected column representations for supported sub expressions. Simple column references and chain of
+        // dereferences on a variable are supported right now.
+        Set<ConnectorExpression> projectedExpressions = projections.stream()
+                .flatMap(expression -> extractSupportedProjectedColumns(expression).stream())
+                .collect(toImmutableSet());
+
+        Map<ConnectorExpression, HiveApplyProjectionUtil.ProjectedColumnRepresentation> columnProjections = projectedExpressions.stream()
+                .collect(toImmutableMap(Function.identity(), HiveApplyProjectionUtil::createProjectedColumnRepresentation));
+
+        IcebergTableHandle icebergTableHandle = (IcebergTableHandle) handle;
+
+        // all references are simple variables
+        if (columnProjections.values().stream().allMatch(HiveApplyProjectionUtil.ProjectedColumnRepresentation::isVariable)) {
+            Set<ColumnHandle> projectedColumns = ImmutableSet.copyOf(assignments.values());
+            if (icebergTableHandle.getProjectedColumns().isPresent()
+                    && icebergTableHandle.getProjectedColumns().get().equals(projectedColumns)) {
+                return Optional.empty();
+            }
+            List<Assignment> assignmentsList = assignments.entrySet().stream()
+                    .map(assignment -> new Assignment(
+                            assignment.getKey(),
+                            assignment.getValue(),
+                            ((IcebergColumnHandle) assignment.getValue()).getType()))
+                    .collect(toImmutableList());
+            IcebergTableHandle newTableHandle = new IcebergTableHandle(
+                    icebergTableHandle.getSchemaName(),
+                    icebergTableHandle.getTableName(),
+                    icebergTableHandle.getTableType(),
+                    icebergTableHandle.getSnapshotId(),
+                    icebergTableHandle.getUnenforcedPredicate(),
+                    icebergTableHandle.getEnforcedPredicate(),
+                    Optional.of(projectedColumns));
+
+            return Optional.of(new ProjectionApplicationResult<>(
+                    newTableHandle,
+                    projections,
+                    assignmentsList,
+                    false));
+        }
+
+        ImmutableMap.Builder<String, Assignment> newAssignments = ImmutableMap.builder();
+        ImmutableMap.Builder<ConnectorExpression, Variable> newVariablesBuilder = ImmutableMap.builder();
+        ImmutableSet.Builder<ColumnHandle> projectedColumnsBuilder = ImmutableSet.builder();
```

Comment:
I had a `find` method but took it out, I'm still a bit confused on the reasoning for it.

From the other thread about it,
> IIRC duplicate avoidance in hive connector was intended to sanitize against the following case throwing duplicate-key error for ImmutableMap. Not sure if this is easily demonstrable in a query, but optimizer doesn't explicitly prevent it from happening.
> Assignments: <"a", CH(a)>, <"a_b", CH(a.b)>
> projections: ["a".b, "a_b"]
> if finding seems expensive, we can just verify that columnhandle is exactly the same and skip.

The Hive applyProjection still adds a new entry into `newAssignments`, `newVariablesBuilder`, and `projectedColumnsBuilder` even in the case where an existing column in the assignments is found. All it does is use the existing column handle and it's name rather than create a new one. I think I must be missing something because all I see that's different is skipping the construction of one, identical, column handle.

Reply:
One thing to note here is that `assignments` is a map from String to ColumnHandle. The `String` is created from `Symbol#getName`. The symbol could've been created using the output of previous iteration of applyProjection.  say during one iteration, we return assignments `"a" -> CH(a), "a_b" -> CH(a.b)` from this method. Then the engine can create symbols `a_3` and `a_b_5` respectively.  Now, let's say for the next iteration, a projection like `"a_3".b` will produce variable named `a_b`, which is different from `a_b_5`. However, as long as `a_b_5` is also rewritten in the same manner, we wouldn't need to worry about having multiple variables for the same logical "subcolumn" in the returned assignments. (Which seems true here as well, so an explicit "find" is overly defensive, and not required.)

For the above testcase with the current code, we'll try to put the same string ("a_b") twice in `newAssignments`. and fail .So we shouldn't use immutable builder. It's okay for newVariablesBuilder as input connector expressions are different, and  also okay for projectedColumnsBuilder as it's a set and IcebergColumnHandle is created deterministically with properly defined equality. Does this make sense? If so, could you add a rule test? 

```
            Variable projectedColumnVariable = new Variable(projectedColumnName, expression.getType());
            Assignment newAssignment = new Assignment(projectedColumnName, projectedColumnHandle, expression.getType());
            newAssignments.put(projectedColumnName, newAssignment);

            newVariablesBuilder.put(expression, projectedColumnVariable);
            projectedColumnsBuilder.add(projectedColumnHandle);
```

Reply:
I took a stab at writing a rule test for this case, but the PlanMatcher didn't like it: https://github.com/trinodb/trino/pull/8129/files#diff-68db78292ccf93ffa9b1831d26475307dc02a4e1ad00b4c1913b26d65fb1f3ecR229-R254

cc: @findepi